### ISSUE TITLE
[ValidatePattern()] regex breaks SyntaxHighlighting

### ISSUE THREAD

**Issue by OCram85** [PowerShell/EditorSyntax#26]
### System Details
- Operating system name and version: `Win 7 Prof 64Bit`
- VS Code version: `1.5.3`
- PowerShell extension version: `0.7.2`
- Output from `$PSVersionTable`:

```
Name                           Value                                                                                                   
----                           -----                                                                                                   
PSVersion                      5.0.10586.117                                                                                           
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                 
BuildVersion                   10.0.10586.117                                                                                          
CLRVersion                     4.0.30319.42000                                                                                         
WSManStackVersion              3.0                                                                                                     
PSRemotingProtocolVersion      2.3                                                                                                     
SerializationVersion           1.1.0.1 
```
### Issue Description

I noticed a syntax highlight issue while using regex in `[ValidatePattern()]`. 
#### Example1:

![2016-10-11 08_24_43-invoke-2fausersync ps1 - visual studio code](https://cloud.githubusercontent.com/assets/13993700/19260724/4bf892f2-8f8c-11e6-868f-0e3eebdf3e59.png)
#### Example2:

![2016-10-11 08_15_56-invoke-2fausersync ps1 - visual studio code](https://cloud.githubusercontent.com/assets/13993700/19260655/99171316-8f8b-11e6-9414-de977462ebad.png)

**Comment by daviwil** [comment]
Thanks for filing the issue here Marco!  We'll take a look.

**Comment by ThubLives** [comment]
Syntax highlighting is weird inside the quotes for regex in general, but I have found that the weirdness can escape the quotes and trash the rest of the code any time a closing parentheses is followed -- even distantly -- inside the quotes with a closing square bracket.  It's as though it's not getting treated as a string at all inside the parameter attribute construct.

Take this code for example:
`function Test { Param( [ValidatePattern('(abcd)[a-d]')] [string]$TestString ) }`

Strictly speaking, all you need as a string to cause breakage is `')]'`, but I don't think it's really valid regex. :)

**Comment by ThubLives** [comment]
My workaround is to put a single-line comment at the end of the line, which seems to reset things:

```
function Test {
  Param(
    [ValidatePattern('(abcd)[a-d]')] # This comment fixes VSCode's syntax highlighting strangeness.
    [string]$TestString
  )
}
```

**Comment by kedia990** [comment]
@ThubLives I stumbled into the issue just today, so your suggested workaround couldn't have come at a better time for me!

That said, a clarification - it's not just a single-line comment, but a single line comment with an apostrophe ( ' ) in it which fixes the weirdness. Remove the "VSCode's" word in your comment to see what I mean.

**Comment by gravejester** [comment]
This is a known bug in how strings are scoped inside param blocks. We are working on a fix :)

**Comment by ThubLives** [comment]
Update:  I'm still getting the same parsing behaviour in vscode 1.23.0 and ms-vscode.powershell 1.7.0.

**Comment by omniomi** [comment]
@ThubLives, this is currently fixed on my fork but I have not submitted the pull request yet. I am soliciting opinions on reddit: https://www.reddit.com/r/PowerShell/comments/8hrcwe/attribute_highlighting_in_vs_codeatomsublime_your/ as it changes the way attributes are displayed in general.

In one or two days I will submit the PR. In the mean  time you can try the current definitions if you like:

1. Download 'powershell.tmLanguage.json' from https://gist.github.com/omniomi/dea8f3a80a8187792c74c1250c602264
2. Open your VS Code install directory followed by resources\app\extensions\powershell\syntaxes (C:\Program Files\Microsoft VS Code\resources\app\extensions\powershell\syntaxes on Windows).
3.  Rename the existing 'powershell.tmLanguage.json' to .old.
4. Copy in the downloaded json file.
5. If VS Code is already running use command palette -> reload window.

**Comment by ThubLives** [comment]
I've tried the powershell.tmLanguage.json file you provided, but it doesn't seem to deal with this problem, at least based on my example code above.  What's more, my comment-based workaround doesn't work anymore, though it was probably likely to break given the kinds of changes that happened.

**Comment by omniomi** [comment]
I could have sworn I tested your case but it's been through a few iterations before I added it to git. I will take a look tonight and make sure it's covered before I submit the PR.

**Comment by ThubLives** [comment]
Okay, so it looks like the `[ValidatePattern('(abcd)[a-d]')]` line is just fine as long as it's not followed by _anything_.  That is to say, the comment (and separating whitespace) is now what is causing the problem.  Even just adding a space at the end of the line breaks things, though not as badly as before.  The same is true for other attributes like `[alias('spanakopita')]` followed by anything at all.

Now I have to fiddle with my favourite theme, "Dark (Visual Studio)" to get it to colour types again.  :/

**Comment by omniomi** [comment]
Gotcha. Currently it's expecting an end of line at the end of the match. I'll change it to a word boundary and that should resolve the issue.

**Comment by omniomi** [comment]
Did some fiddling and it's easy enough to fix, I'll add it to my repo tonight along with a few other changes.

![comments](https://user-images.githubusercontent.com/12937335/39779885-ba228c74-52c7-11e8-96da-9e8ea83e1b83.png)

I'll let you know when there's an artifact to test with.

**Comment by ThubLives** [comment]
That's great!  It's nice to see some good on this.

I do see a regression with indexes though.  Numbers, strings, and variable names are all getting coloured the same as types.  I don't know what other basic types I'm missing, but I'm guessing they would behave the same.

`$Array[$Variable] = $OtherArray[$Variable - 1] + $HashTable['spanakopita']`

**Comment by omniomi** [comment]
@ThubLives can you give this guy a try: https://gist.github.com/omniomi/dea8f3a80a8187792c74c1250c602264 should address comments/whitespace on the same line as attributes and also fix indices following variables.

I also re-scoped a couple things to be more cosmetically pleasing in the majority of themes. Unfortunately the "right" scopes aren't always covered by theme makers if they cover things not common to most languages.

**Comment by ThubLives** [comment]
I dropped in the updated syntax file and visually scanned a few of my more complicated scripts and try as I might, can't find anything to complain about.  Great work!

**Comment by omniomi** [comment]
Fixed by #94
